RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -207,7 +207,87 @@ namespace {
     }
     return tofpid;
   }
+
+  bool getTrajectoryStateClosestToBeamLine(const Trajectory& traj,
+                                           const reco::BeamSpot& bs,
+                                           const Propagator* thePropagator,
+                                           TrajectoryStateClosestToBeamLine& tscbl) {
+    // get the state closest to the beamline
+    TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface =
+        traj.closestMeasurement(GlobalPoint(bs.x0(), bs.y0(), bs.z0())).updatedState();
+
+    if (!stateForProjectionToBeamLineOnSurface.isValid()) {
+      edm::LogError("CannotPropagateToBeamLine") << "the state on the closest measurement isnot valid. skipping track.";
+      return false;
+    }
+
+    const FreeTrajectoryState& stateForProjectionToBeamLine = *stateForProjectionToBeamLineOnSurface.freeState();
+
+    TSCBLBuilderWithPropagator tscblBuilder(*thePropagator);
+    tscbl = tscblBuilder(stateForProjectionToBeamLine, bs);
+
+    return tscbl.isValid();
+  }
+
+  bool trackPathLength(const Trajectory& traj,
+                       const TrajectoryStateClosestToBeamLine& tscbl,
+                       const Propagator* thePropagator,
+                       double& pathlength) {
+    pathlength = 0.;
+
+    bool validpropagation = true;
+    double pathlength1 = 0.;
+    double pathlength2 = 0.;
+
+    //add pathlength layer by layer
+    for (auto it = traj.measurements().begin(); it != traj.measurements().end() - 1; ++it) {
+      const auto& propresult = thePropagator->propagateWithPath(it->updatedState(), (it + 1)->updatedState().surface());
+      double layerpathlength = std::abs(propresult.second);
+      if (layerpathlength == 0.) {
+        validpropagation = false;
```

Comment:
is it useful to continue the propagation if it's already known to be invalid?
(if invalid results aren't used, this function could just exit here, and skip some downstream computation)

Reply:
@pmeridian @lgray 
do you think it would be OK to skip and return false (I suppose this is a question to the "add distance from bs to first measurement" below as well, although there it's much less likely to happen) ?
I can change it in this PR, but not sure if there was some implied need to have a partial path length

Reply:
after adding some debugging, I see that this is happening quire often

Reply:
"quite often" was an overstatement. It's affecting about 0.4% of the calls to trackPathLength.
The cost of all trackPathLength calls is about 8% of the TrackExtenderWithMTDT time.

Reply:
Okay, so this is < 0.3% of the total cost, probably not worth optimizing.

Reply:
I think that it's in the ballpark of 0.001% of reco time (0.4%*8%*2.5%).
I would've implemented it if I'd see a unique solution.
- The call in ::buildTrack is protected and can return early without changes in results
- The call in the main loop over tracks (was in ::find_hits_in_dets) had no protection and can lead to different content in the matched hit info. 
    - I made a test with return at every validpropagation false and get about 20 tracks with differences (out of 360K tracks).

So, I'd rater leave this change out of this PR, which aims to be purely technical (produce the same output)